src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ExhaustivenessComputer.java

```diff
@@ -229,16 +228,26 @@ private Set<PatternDescription> reduceBindingPatterns(Type selectorType, Set<Pat
 
                             return instantiated != null && types.isCastable(selectorType, instantiated);
                         });
+                        Set<Symbol> filteredPermitted = new HashSet<>(permitted);
 
                         for (PatternDescription pdOther : patterns) {
                             if (pdOther instanceof BindingPattern bpOther) {
-                                Set<Symbol> currentPermittedSubTypes =
-                                        allPermittedSubTypes(bpOther.type.tsym, s -> true);
+                                Set<Symbol> currentSubTypes;
+
+                                if (bpOther.type.tsym.isAbstract()) {
+                                    currentSubTypes =
+                                            permitted.stream()
+                                                     .filter(perm -> types.isSubtype(types.erasure(perm.type),
```

Comment:
Can we also add a test with a generic hierarchy to witness the alternative path in the `types.erasure` part? I am sure the following would pass thought so feel free to ignore:

```
class SomeType {}
sealed interface Base<T extends SomeType> permits Special, Value {}
non-sealed interface Value<T extends SomeType>  extends Base<T> {}
sealed interface Special<T extends SomeType> extends Base<T> permits SpecialValue {}
non-sealed interface SpecialValue<T extends SomeType> extends Value<T>, Special<T> {}

static <T extends SomeType> int demo(final Base<T> base) {
    return switch (base) {  
         case Value<T> value -> 0;
    };
}
```